### ISSUE TITLE
add more runtime metrics to activation records?

### ISSUE THREAD

**Issue by jthomas** [apache/incubator-openwhisk#3224]
This is an open-ended question about the possibility of adding more runtime metrics to the activation records, like #2846 did.

I've been working on a [metrics library](https://github.com/jthomas/openwhisk-metrics) for openwhisk that records runtime metrics using a Node.js function proxy. If is was possible for the platform to include these values directly in the activation records, it would negate the need for a separate library and support all runtimes.

This would be dependent on the values exposed by [docker stats API](https://docs.docker.com/engine/api/v1.33/#operation/ContainerStats)? 

Metric values that would be useful for users include...
 - Cold start (I know this can be inferred from missing `initTime` but might be nice to have a convenience value).
- Memory usage including OOM errors?
- CPU Usage

**Comment by rabbah** [comment]
OOM errors are recorded in the activation record - this is inferred from `docker inspect`; whatever we can infer that's sensible 👍 to including in the activation.

**Comment by ScottChapman** [comment]
it would be useful to have an `elapsedTime` which is how much time is spent in the actual function. From this is it possible for one to determine what the "overhead" is.

**Comment by rabbah** [comment]
That’s the “duration”. Yes there’s a small/epsilon difference vs measuring right around main. Have you observed a big deviation between that and duration?

**Comment by ScottChapman** [comment]
I have, though to be fair what I have to do invoke a trigger with my transaction and wait for the result before my action can finish, so I can't really include that.

So, I delta is a combo cost of sending an event to a trigger plus that small epsilon.

In some cases it is very small, but in other cases it is 0.5 - 1.5 seconds

**Comment by rabbah** [comment]
I don’t see how we measure elapsed time without including time to resolve the promise/asynchronous activity. Do you have a generic code snippet with timers added?

**Comment by ScottChapman** [comment]
Yea, but the point is I don't have to invoke the trigger if I am not doing my own measurements. So really, there is no good way for me to measure myself without impacting the results in some fashion.

Generically what I do in my code is something line:
```
var start = Date.now();
...
...
await ow.trigger.invoke({
  name: "Transaction",
  params: {
    start: start,
    end: Date.now()
  }
 })
return {status: "OK!"}
```
I need to do the await on the invocation otherwise the action might terminate before it has had a chance to post. Though it is an unblocking post.
